No longer generate extra atoms to compare generic types

kyleheadley opened the pull request:
Originally, to compare function outputs with other constraints, we used the function's CV, which may have been generic. This made comparisons tricky, so extra atoms were added to fill out the type, but it resulted in over-constrained, no-longer-generic functions.

Now we use the type variable data already being collected to determine function output in a context-sensitive way. Most of this PR is about propagating that existing type variable information to other parts of the code-base.

kyleheadley commented on clang/lib/3C/ConstraintResolver.cpp:
```diff
@@ -403,6 +423,10 @@ CSetBkeyPair ConstraintResolver::getExprConstraintVars(Expr *E) {
       CVarSet ReturnCVs;
       BKeySet ReturnBSet = EmptyBSet;
 
+      ProgramInfo::CallTypeParamBindingsT TypeVars;
```
What is duplicated?

aaronjeline replied:
Creating a set, checking `hasTypeParamBindings` and then using the `getTypeParamBindings` to fill the set (leaving it empty if the check is false). You also do this at line 219 in `ConstraintBuilder.cpp`. Wrapping it in a method inside `ProgramInfo` and giving it a short comment might make things easier for a future reader

aaronjeline replied:
Creating a set, checking hasTypeParamBindings and then using the getTypeParamBindings to fill the set (leaving it empty if the check is false). You also do this at line 219 in ConstraintBuilder.cpp. Wrapping it in a method inside ProgramInfo and giving it a short comment might make things easier for a future reader

kyleheadley replied:
I just tried and I can't do it unless you know how. A returned reference can't be null, so this interface requires a branch with check and get, which is what the code here and elsewhere uses. A pointer interface won't simplify the code except to replace a function call with a null check, which could be forgotten. Saving a field with a 'zero' element that could be referenced would make this more algebraic, but might look strange to someone unfamiliar with the technique.

How should I proceed?

aaronjeline replied:
```
void ProgramInfo::getTypeVars(ProgramInfo::CallTypeParamBindingsT& TyVars, CallExpr *CE, ASTContext *C) const {
  if (hasTypeParamBindings(CE, C))
    TyVars = getTypeParamBindings(CE, C);
}
```
I guess this is the best you can do. Up to you if you think this is worse. I tend to err towards more names for things but you only copy it twice so up to you which you think if clearer.

kyleheadley replied:
You still need to initialize `TyVars` and call the function, so you save one line of code and add complexity. I think it's fine, but not worth making the change, if you don't mind.

aaronjeline commented on clang/test/3C/root_cause.c:
```diff
@@ -33,7 +33,7 @@ void test1() {
   c++; // expected-warning {{1 unchecked pointer: Pointer arithmetic performed on a function pointer}}
 
-  int *d = malloc(1); // expected-warning {{1 unchecked pointer: Unsafe call to allocator function}}
+  int *d = malloc(1); // expected-warning {{1 unchecked pointers: Unsafe call to allocator function}}
```
```suggestion
  int *d = malloc(1); // expected-warning {{1 unchecked pointer: Unsafe call to allocator function}}
```
Small test change to get test passing

aaronjeline reviewed (commented):
One more quick thing I noticed

mattmccutchen-cci commented on clang/test/3C/hash.c:
```diff
@@ -9,10 +9,7 @@
-                                          : itype(_Array_ptr<T>)
-                                                byte_count(size),
-                                            unsigned int size)
+_Itype_for_any(T) void vsf_sysutil_memclr(void *p_dest : itype(_Array_ptr<T>) byte_count(size), unsigned int size)
```
Nit: Please change the formatting back or find some other way to keep the line length under 80 characters, as [Microsoft asked](https://github.com/microsoft/checkedc-clang/pull/993#issuecomment-791743175), even though it looks awkward.  Otherwise, I'll have to do it when I prepare the next omnibus PR to Microsoft.  I don't mind doing it, but I thought I should make you aware of the issue so maybe we can avoid changing the formatting back and forth unnecessarily.

mattmccutchen-cci replied:
> As I recall, this function was originally rewritten (unnecessarily) onto one line. After the change, it is no longer rewritten and fails the unmodified check line.

Great!

> Is the solution to add multiple checks for the one function?

Yes, or more specifically, `CHECK-NEXT`.  Here's an example:
https://github.com/correctcomputation/checkedc-clang/blob/cf35b121de3211cbcac3e107edacaaf8c89e45e0/clang/test/3C/simple_locals.c#L18-L20

kyleheadley commented:
This should be ready for review again, I believe I've addressed all the comments above. Let me know if there are still changes requested or if I'm good to merge.

mattmccutchen-cci commented on clang/test/3C/hash.c:
```diff
@@ -9,8 +9,14 @@
+                                                byte_count(size),
+                                            unsigned int size)
+// CHECK_ALL: _Itype_for_any(T) void vsf_sysutil_memclr(void *p_dest
+// CHECK-NEXT: : itype(_Array_ptr<T>)
+// CHECK-NEXT: byte_count(size),
+// CHECK-NEXT: unsigned int size)
```
Ah, I missed that you're using a custom check prefix, so I think this should be:
```suggestion
// CHECK_ALL-NEXT: : itype(_Array_ptr<T>)
// CHECK_ALL-NEXT: byte_count(size),
// CHECK_ALL-NEXT: unsigned int size)
```
Example:
https://github.com/correctcomputation/checkedc-clang/blob/cf35b121de3211cbcac3e107edacaaf8c89e45e0/clang/test/3C/stmtExprA.c#L16-L20

mattmccutchen-cci reviewed (dismissed):
Just a nit.

kyleheadley commented:
@aaronjeline I believe you still have an outstanding request for changes. Was there something else you asked me to do here or are we good to merge?

aaronjeline commented:
Good to merge
